plan/preprocess.go

```diff
@@ -561,3 +598,82 @@ func (p *preprocessor) resolveAlterTableStmt(node *ast.AlterTableStmt) {
 		}
 	}
 }
+
+// handleTableSources checks name duplication
+// and puts the table source in current resolverContext.
+// Note:
+// "select * from t as a join (select 1) as a;" is not duplicate.
+// "select * from t as a join t as a;" is duplicate.
+// "select * from (select 1) as a join (select 1) as a;" is duplicate.
+func (p *preprocessor) handleTableSource(ts *ast.TableSource) {
+	for _, v := range ts.GetResultFields() {
+		v.TableAsName = ts.AsName
+	}
+	switch ts.Source.(type) {
+	case *ast.TableName:
+		var name string
+		if ts.AsName.L != "" {
+			name = ts.AsName.L
+		} else {
+			tableName := ts.Source.(*ast.TableName)
+			name = p.tableUniqueName(tableName.Schema, tableName.Name)
+		}
+		if _, ok := p.tableMap[name]; ok {
+			p.err = errors.Errorf("duplicated table/alias name %s", name)
+			return
+		}
+		p.tableMap[name] = len(p.tables)
+	case *ast.SelectStmt:
+		name := ts.AsName.L
+		if _, ok := p.derivedTableMap[name]; ok {
+			p.err = errors.Errorf("duplicated table/alias name %s", name)
+			return
+		}
+		p.derivedTableMap[name] = len(p.tables)
+	}
+	dupNames := make(map[string]struct{}, len(ts.GetResultFields()))
+	for _, f := range ts.GetResultFields() {
+		name := f.ColumnAsName.L
+		if name == "" {
+			name = f.Column.Name.L
+		}
+	}
+	for _, f := range ts.GetResultFields() {
+		// duplicate column name in one table is not allowed.
+		// "select * from (select 1, 1) as a;" is duplicate.
+		name := f.ColumnAsName.L
+		if name == "" {
+			name = f.Column.Name.L
+		}
+		if _, ok := dupNames[name]; ok {
+			p.err = errors.Errorf("Duplicate column name '%s'", name)
+			return
+		}
+		dupNames[name] = struct{}{}
+	}
+	p.tables = append(p.tables, ts)
+}
+
+func (p *preprocessor) handleUnionSelectList(u *ast.UnionSelectList) {
+	firstSelFields := u.Selects[0].GetResultFields()
```

Comment:
Why do we need to change result fields. It seems we don't depend on it any more.